birdhouse/components/thredds/config/magpie/providers.cfg.template

```diff
@@ -33,6 +33,7 @@ providers:
           dodsC,
           wcs,
           wms,
-          ncss,
+          ncss/grid,
+          ncss/point,
```

Comment:
Did you test if this works?

I do not know if Magpie/Twitcher handles this properly.
The data types were not planned to include a `/`, so I'm not sure if a `url.split("/")` might happen somewhere in the code leading to misinterpretation of the targeted `data_type`.

Reply:
Ok maybe we can't change the path... the URL construction docs for this service look like point or graph needs to be a subpath under ncss:

https://docs.unidata.ucar.edu/tds/current/userguide/netcdf_subset_service_ref.html#url-construction

Reply:
> The data types were not planned to include a /, so I'm not sure if a url.split("/") might happen somewhere in the code leading to misinterpretation of the targeted data_type.

Yup magpie splits it up (see [here](https://github.com/Ouranosinc/Magpie/blob/dc0a9a9871f13f1c55b219d644245e9cc3e1c96e/magpie/services.py#L1372)) so our only option at this point is to keep the magpie config as it is and treat `grid/` and `point/` as directory paths if we want to differentiate their magpie permissions.

That's going to cause other confusion though... 

Right now we have URL paths like: 

- `${FQDN}/twitcher/ows/proxy/thredds/ncss/datasets/...`
- `${FQDN}/twitcher/ows/proxy/thredds/dodsC/datasets/...`

So if we want to set specific permissions on the `datasets/` directory we can do that by setting one directory permission rule on the `datasets/` subdirectory and it will work for all services.

But if we instead have:

- `${FQDN}/twitcher/ows/proxy/thredds/ncss/point/datasets/...`
- `${FQDN}/twitcher/ows/proxy/thredds/dodsC/datasets/...`

Then we would need to set the same permission rule on `datasets/` as well as `point/datasets/`. If we don't set the same rule on `point/datasets/` then the rule won't apply to resources accessed by the `ncss/point` service.

Please let me know if I'm interpreting this correctly @fmigneault 

If I'm right then I think we need to update Magpie to handle this use case.

Reply:
I'm going to revert my changes for now until we figure something out.

Reply:
> I'm going to revert my changes for now until we figure something out.

Yes because in the current state (just `ncss` in `magpie/providers.cfg.template`) it already works in my test server.

I have a feeling Magpie will just let through everything after `/thredds/ncss/...` in the URL so `/thredds/ncss/grid/` and `/thredds/ncss/point/` just work.  Also this future-proof us from `/thredds/ncss/<another_split>/`.

However if we ever need to have different Allow/Deny permissions for `nccs/grid` or `ncss/point`, then we need to distinguish between them.

Reply:
Thanks for validating. 
Yes, you're interpreting it correctly. 

You would need to duplicate the entire permission hierarchy 3 times for anything under "datasets", because the dirs for THREDDS would be defined as follows.  

````
birdhouse
  datasets
    ...  # current ncss and others
  points
    datasets
  grid
    datasets
```


This also causes a weird situation where an actual dir named "points" or "grid" under "birdhouse" (eg: `${FQDN}/twitcher/ows/proxy/thredds/dodsC/points`) would have a different set of permissions that might overlap/conflict or be irrelevant to what points/grid "datasets" define.

Reply:
Alright that makes sense. We should probably make an issue for this over on the Magpie repo. Do you want to do that or shall I?

Reply:
> You would need to duplicate the entire permission hierarchy 3 times for anything under "datasets", because the dirs for THREDDS would be defined as follows.
> 
> ```
> birdhouse
>   datasets
>     ...  # current ncss and others
>   points
>     datasets
>   grid
>     datasets
> ```
> 
> This also causes a weird situation where an actual dir named "points" or "grid" under "birdhouse" (eg: `${FQDN}/twitcher/ows/proxy/thredds/dodsC/points`) would have a different set of permissions that might overlap/conflict or be irrelevant to what points/grid "datasets" define.

I am completely lost here.  Does this means we have a problem with the current Magpie config or not?

Reply:
@mishaschwartz Yes, you can create the issue.
@tlvu Magpie cannot handle this case currently because it wasn't designed for this. It assumes THREDDS "services" don't contain a `/`, so it could rely on `{Twitcher-prefix}/{THREDDS_name}/{THREDDS_service}/{rest-as-dir/file}`.

Reply:
> @tlvu Magpie cannot handle this case currently because it wasn't designed for this. It assumes THREDDS "services" don't contain a `/`, so it could rely on `{Twitcher-prefix}/{THREDDS_name}/{THREDDS_service}/{rest-as-dir/file}`.

@fmigneault I am still lost.  What case are you talking about?

Just to be clear, in the current state, with Misha's proposed Magpie config change rollback, do we have any problems?

Been running some Jenkins on the new Thredds and I am having some weird errors.  Not sure if it's my test system or the code.  So if you foresee some problem, please let me know.

Reply:
The `ncssGrid` and `ncssPoint` services won't be protected like other services do (ie: a common browse/read/write set for all services) since the paths are not handled.
For the moment, if you want them protected, you need to duplicate the permission hierarchy.

i.e.: 

A file accessed by `ncssGrid` via `${FQDN}/twitcher/ows/proxy/thredds/ncss/point/datasets/sub/file.nc` will actually go through `ncss` service, and `point` will be considered by Magpie as if it was any other directory (like `datasets` or `sub`).

Therefore, the same `datasets/sub/file.nc` would need 3 sets of permissions:

- datasets/sub/file.nc  => (browse,read,write) for anything currently handled
- grid/datasets/sub/file.nc  => (read) only if accessing via ncssGrid
- point/datasets/sub/file.nc  => (read) only if accessing via ncssPoint

and those (read) need to be duplicated for every user/group/dir/file combination applicable

Reply:
> The `ncssGrid` and `ncssPoint` services won't be protected like other services do (ie: a common browse/read/write set for all services) since the paths are not handled.

@fmigneault I am surprised.  If we "Deny or Allow" everything under `ncss` path, isn't both of those variant `grid/datasets/sub/file.nc` and `point/datasets/sub/file.nc` will be properly denied or allowed, because they are all under `ncss/` path?

Under `ncss` is read only no write anyways, since there is no write for that kind of path.

Reply:
@tlvu 
Yes. 
A deny top-level will restrict everything lower. But usually, a deny is "undone" by an allow on a lower specific resource we want to grant access. The same issue also happens even when working the other way around -- allow recursive followed by restricted deny lower (see below example).

With the current implementation of THREDDS that checks the service [accses-node] (dodC, ncss, etc.) between Twitcher-proxy-path and the rest of the dir/file path, those resources would be at different "level" for the same file reference from the point of view of Magpie. So you would need to duplicate your 'allows'/'denies' across [accses-node].

```
RESOURCE        PERMISSION APPLIED / RESULT
===========================================
THREDDS         allow-recursive

ncss            [access-mode] (not a resource, cannot have permissions directly)
  datasets      deny-recursive
     sub        allow-recursve (all but sub's contents are blocked)

  public        allow-recursive
     secure     deny-recursive
       secret   allow-match only for user-1

  point         <-- THREDDS sees this as another [access-mode], but Magpie thinks its a RESOURCE!
     datasets   
       sub      <==== OH! OH! undesired full open-access (it's not under the denied "/datasets")

     public     allow-recursive (but because on THREDDS, not the one on "/dataset")
       secure   if you forget to replicate above 'secure', this is still full open acces

  grid          (repeat again, another set of duplicat permissions, for each group/user combination)
    ...
```

Reply:
> A deny top-level will restrict everything lower. But usually, a deny is "undone" by an allow on a lower specific resource we want to grant access. The same issue also happens even when working the other way around -- allow recursive followed by restricted deny lower (see below example).

Oh OK !!! Now I understand, thanks !  Top-level Allow or Deny will work.  But adding an exception under the top-level will not.